File: core/src/server/model/UpgradeDefinition.scala
package server.model

import java.util.ArrayList
import java.util.HashMap
import java.awt.image.BufferedImage
import java.awt.Graphics
import engine.general.utility.Location
import scala.collection.JavaConversions._
import engine.rts.model.Resource

object UpgradeDefinition{
    var upgradeList=new ArrayList[UpgradeDefinition]()
    def getUpgrade(keyPress:Integer):UpgradeDefinition={
        for (upgrade<-upgradeList){
            if (upgrade.getBuildKey()==keyPress){
                return upgrade
            }
        }
        return null
    }
}


/**
 * This class will store the data for each upgrade type after reading from the file.
 * @param improvementMap The data that was read from the JSON file.
 * @param resourceList The list of resources in the game.
 */
class UpgradeDefinition(improvementMap:Map[Any,Any],resourceList:ArrayList[Resource]) extends java.io.Serializable{

    val name=improvementMap.get("name").get.asInstanceOf[String]//Set the name
    val imageName=improvementMap.get("image").get.asInstanceOf[String] //Set the name
    var resourceCosts=new HashMap[String,Integer]() //This represents how much an upgrade costs.
    var resourceBonus=new HashMap[String,Double]()//This represents the resource income setBonus.

    for(resource<-resourceList){
        val resName:String=resource.getName()
        resourceCosts.put(resName,(improvementMap.get(resName+"_cost").get.asInstanceOf[String]).toInt)

        if(improvementMap.contains(resName+"_bonus")){
            resourceBonus.put(resName,improvementMap.get(resName+"_bonus").get.asInstanceOf[String].toDouble)
        }
        else{
            resourceBonus.put(resName,1.0)
        }
    }

    val  moveBonus=improvementMap.get("move_bonus").get.asInstanceOf[String].toDouble
    val description=improvementMap.get("descr").get.asInstanceOf[String]

    //Set more of the attributes.
    val attackPower=improvementMap.get("attack_power").get.asInstanceOf[String]
    val attackBonus=improvementMap.get("attack_bonus").get.asInstanceOf[String].toDouble
    val defenseBonus=improvementMap.get("defense_bonus").get.asInstanceOf[String].toDouble
    val keyTemp=improvementMap.get("build_key").get.asInstanceOf[String].toCharArray()
    val buildKey=keyTemp(0)

    //Add the upgrade to the list of upgrades.
    val id=UpgradeDefinition.upgradeList.size()
    UpgradeDefinition.upgradeList.add(this)//Add improvement to list of upgrades.
    val myImage: BufferedImage = null //This represents the image

    def getInfo=description
    def getMoveBonus=moveBonus
    def getID():Integer=id
    def getImageLocation():String=imageName

    //TODO:Make sure different improvements aren't drawn in same location
    def draw (g: Graphics, loc: Location){
        val dx: Int = loc.getX.asInstanceOf[Int]
        val dy: Int = loc.getY.asInstanceOf[Int] - 15
        g.drawImage(myImage, dx - 40, dy - 10, 30, 30, null)
    }

    def getResourceBonus(res:String)=resourceBonus.get(res)
    def getAttackStr=attackPower.toInt
    def getResourceBonus():HashMap[String,Double]=resourceBonus
    def getCost():HashMap[String,Integer]=resourceCosts
    def getBuildKey():Integer=buildKey

    def costStr():String={
        var s="Cost:"
        for ((name,cost)<-resourceCosts){
            s+=","+cost
        }
        return s
    }
    
    def printData(){
        println(name+" "+imageName+" costs:"+resourceCosts+" bonuses:"+resourceBonus+" attackBonus:"+attackBonus+" attackPower:"+attackPower+" defenseBonus:"+defenseBonus)
    }
}

File: core/src/server/model/UpgradeDefinition.java
package server.model;


import java.util.ArrayList;
import java.util.HashMap;
import java.awt.Graphics;

import com.badlogic.gdx.utils.JsonValue;

import engine.general.utility.Location;
import engine.rts.model.Resource;

public class UpgradeDefinition implements java.io.Serializable{

	public static class UpgardeFactory{
		public ArrayList<Resource> resourceList=new ArrayList<Resource>();
	}
	
	public static ArrayList<UpgradeDefinition>upgradeList=new ArrayList<UpgradeDefinition>();
    public static void setResourceList(ArrayList<UpgradeDefinition> u){
    	upgradeList=u;
    }
    
	public static UpgradeDefinition getUpgrade(int keyPress){
        for (UpgradeDefinition upgrade:upgradeList){
            if (upgrade.getBuildKey()==keyPress){
                return upgrade;
            }
        }
        return null;
    }
	
	final String name;
	final String imageName;
	final String description;
	
	final double moveBonus;
	final String attackPower;
	public final double attackBonus;
	public final double defenseBonus;
	final char buildKey;
	
	HashMap<String,Integer> resourceCosts=new HashMap<String,Integer>();
	HashMap<String,Double> resourceBonus=new HashMap<String,Double>();
	
	final int id;
	
	public UpgradeDefinition(JsonValue data,ArrayList<Resource> resourceList){
		name=data.getString("name");
		imageName=data.getString("image");
		for(Resource resource:resourceList){
			String resName=resource.getName();
			resourceCosts.put(resName, Integer.parseInt(data.getString(resName+"_cost")));
			if(data.hasChild(resName+"_bonus")){
				resourceBonus.put(resName,Double.parseDouble(data.getString(resName+"_cost")));
			}else{
				resourceBonus.put(resName,1.0);
			}
		}
		moveBonus=data.getDouble("move_bonus");
		description=data.getString("descr");
		attackPower=data.getString("attack_power");
		attackBonus=data.getDouble("attack_bonus");
		defenseBonus=data.getDouble("defense_bonus");
		buildKey=data.getChar("build_key");
	
		//Add the upgrade to the list of upgrades.
		id=upgradeList.size();
		upgradeList.add(this);//Add improvement to list of upgrades.
	}
	
	public String getName(){
		return name;
	}
	public String getInfo(){
		return description;
	}

	public double getMoveBonus(){
		return moveBonus;
	}
	
	public int getID(){
		return id;
	}
	
	public String getImageLocation(){
		return imageName;
	}


    //TODO:Refactor code so that this method is unecessary
    public void draw (Graphics g,Location loc){
        System.err.println("This method not yet implemented");
    }

    public double getResourceBonus(String res){
    	return resourceBonus.get(res);
    }
    
    public int getAttackStr(){
    	return Integer.parseInt(attackPower);
    }
    
    public HashMap<String,Double> getResourceBonus(){
    	return resourceBonus;
    }
    
    public HashMap<String,Integer >getCost(){
    	return resourceCosts;
    }
    
    public Integer getBuildKey(){
    	return (int)buildKey;
    }

    public String costStr(){
        String s="Cost:";
        for (String name:resourceCosts.keySet()){
            s+=","+resourceCosts.get(name);
        }
        return s;
    }
    
    public void printData(){
        System.out.println(name+" "+imageName+" costs:"+resourceCosts+" bonuses:"+resourceBonus+" attackBonus:"+attackBonus+" attackPower:"+attackPower+" defenseBonus:"+defenseBonus);
    }
}
